Fix #14 - Support all regions via Regions API

barbeau opened the pull request:
~~*Please don't merge*~~

~~Initial work in progress for improving region discovery.  Note that I modified the `onebusaway-client-library` to include some new functionality for finding only "nearby" regions (https://github.com/OneBusAway/onebusaway-client-library/issues/7), so you'll need to update your local Maven repo for this to compile.~~

~~Note that I haven't actually tested this yet, but most of the infrastructure should now be in place.~~

See below - PR has been updated and ready to merge.

barbeau commented:
Alright, I've tested this and it works with Tampa stops now as well.  I have a few more items of cleanup to do and then I'll ping again when it's ready for review.

barbeau commented:
@philipmw Ok, this PR is review for review/merge.  I've squashed the commits and force-pushed the merge-ready commit into this same PR.

philipmw commented on src/main/java/org/onebusaway/alexa/lib/ObaUserClient.java:
```diff
@@ -35,18 +37,37 @@
+     * @return the closest region in the Regions API to the given location, or null if there are no nearby regions
+     * (within 100 miles of the provided location) or a region couldn't be found.
+     */
+    public static ObaRegion getClosestRegion(Optional<Location> l) {
```
Since this function requires a non-null `Location` argument (Optional.empty is not supported), it should take `Location` directly, leaving to the caller the matter of what should happen if there's no location.

philipmw replied:
Since `getClosestRegion` can return null, we should wrap the return type with `Optional<>`.

philipmw commented on src/main/java/org/onebusaway/alexa/lib/ObaUserClient.java:
```diff
@@ -35,18 +37,37 @@
+     */
+    public static ObaRegion getClosestRegion(Optional<Location> l) {
+        log.debug("Invoked getClosestRegion() with location " + l.toString());
+        boolean enforceProximityThreshold = true;
```
Is this just for documentation purposes? If so, I slightly prefer:

    return RegionUtils.getClosestRegion(
      regions,
      l,
      true); //enforce proximity threshold

but w/e.

philipmw commented on src/main/java/org/onebusaway/alexa/lib/ObaUserClient.java:
```diff
@@ -35,18 +37,37 @@
 
-    public ObaUserClient(String endpoint) {
-        log.debug("Instantiating ObaUserClient with endpoint " + endpoint);
+    public ObaUserClient(String obaBaseUrl) {
```
You're passing into this method, as an argument, directly the return value of `getObaBaseUrl()`. That currently returns either String or null, leading to an occasional error or crash.

philipmw commented on src/main/java/org/onebusaway/alexa/AuthedSpeechlet.java:
```diff
@@ -35,17 +34,13 @@
         this.userData = userData;
-        this.obaUserClient = new ObaUserClient(
-                obaAgencies.agencyForCity(userData.getCity()).get());
+        this.obaUserClient = new ObaUserClient(userData.getObaBaseUrl());
```
In combination with my other comments -- here's where we'll need to decide how to handle `userData.getObaBaseUrl()` returning null, before instantiating `ObaUserClient`.

philipmw commented:
Overall I am delighted to have this PR! I am glad we are able to remove the hacky `ObaAgencies`.

barbeau commented:
@philipmw Ok, I believe I addressed all the feedback.  I'm new to Java's Optional<> - fun :).

As far as handling a null value for the `obaBaseUrl` - I've updated the flow to ask for the city again in this case, so we should never persist a region without this information.  tl;dr - for now we can assume this shouldn't happen (this is a required field in the Regions API for the apps, currently).  Please take a look and make sure this corresponds with your thoughts on this.  I'll also think about how we might need to change the oba-client-library to support future cases that could arise.

philipmw commented on src/main/java/org/onebusaway/alexa/lib/ObaUserClient.java:
```diff
@@ -35,18 +37,39 @@
+        log.debug("Invoked getClosestRegion() with location " + l.toString());
+        ObaRegionsResponse response = ObaRegionsRequest.newRequest().call();
+        ArrayList<ObaRegion> regions = new ArrayList<>(Arrays.asList(response.getRegions()));
+        return Optional.of(RegionUtils.getClosestRegion(
```
Since RegionUtils.getClosestRegion can return null, you need `Optional.ofNullable` instead of `Optional.of`. `of` will raise an exception if you give it null.

philipmw commented on src/main/java/org/onebusaway/alexa/lib/ObaUserClient.java:
```diff
@@ -35,18 +37,39 @@
+            ObaApi.getDefaultContext().setBaseUrl(obaBaseUrl);
         } catch (URISyntaxException e) {
-            log.error("OBA Endpoint we constructed was invalid: " + e.getMessage());
+            log.error("ObaBaseUrl we constructed was invalid: " + e.getMessage());
```
Since you are here already, can you add a `throw e` for this exception?

barbeau commented:
Alright, updated.  Throwing the uri exception cascaded several levels and I wasn't sure the best way to finally handle it in the Speechlet - check it out and let me know what you think.  I updated this to also use the test framework from your last commit, added a Tampa test as well.
